client_plugin/drivers/shared/shared_driver.go

```diff
@@ -350,8 +352,87 @@ func (d *VolumeDriver) Create(r volume.Request) volume.Response {
 // Remove - removes individual volume. Docker would call it only if is not using it anymore
 func (d *VolumeDriver) Remove(r volume.Request) volume.Response {
 	log.WithFields(log.Fields{"name": r.Name}).Info("Removing volume ")
+	var msg string
+	var volRecord VolumeMetadata
+
+	// Cannot remove volumes till plugin completely initializes
+	// because we don't know if it is being used or not
+	if d.RefCounts.IsInitialized() != true {
+		msg = fmt.Sprintf(plugin_utils.PluginInitError+" Cannot remove volume %s",
+			r.Name)
+		log.Error(msg)
+		return volume.Response{Err: msg}
+	}
+
+	// Docker is supposed to block 'remove' command if the volume is used.
+	// Check on the local refcount first
+	if d.GetRefCount(r.Name) != 0 {
+		msg = fmt.Sprintf("Remove failed: Containers on this host VM are still using volume %s.",
+			r.Name)
+		log.Error(msg)
+		return volume.Response{Err: msg}
+	}
+
+	// Check status, then global refcount
+	keys := []string{
+		etcdPrefixState + r.Name,
+		etcdPrefixGRef + r.Name,
+		etcdPrefixInfo + r.Name,
+	}
+	entries, err := d.readVolMetadata(keys)
+	if err != nil {
+		msg = fmt.Sprintf("Remove failed: Failed to read metadata for volume %s from KV store. %v",
+			r.Name, err)
+		log.Warningf(msg)
+		return volume.Response{Err: msg}
+	}
+	// Unmarshal Info key
+	err = json.Unmarshal([]byte(entries[InfoIdx].value), &volRecord)
+	if err != nil {
+		msg = fmt.Sprintf("Remove failed: Failed to unmarshal data. %v", err)
+		log.Warningf(msg)
+		return volume.Response{Err: msg}
+	}
+	// Every other state can be handled by global refcount value.
+	// Only check for Intermediate state as that could mean grefc wont
+	// remain 0 for very long.
+	if volStatus(entries[stateIdx].value) == volStateIntermediate {
+		msg = fmt.Sprintf("Remove failed: Volume metadata is being updated. Please try later.")
+		log.Warningf(msg)
+		return volume.Response{Err: msg}
+	}
+	grefc, _ := strconv.Atoi(entries[GRefIdx].value)
+	if grefc != 0 {
```

Comment:
If there is no connectivity to etcd, readVolMetadata()  will fail.

If this docker host is recovering from a crash, and we want to read from etcd:
1. **By talking to this node:** etcd will throw an error as it will not be ready. While recovering we delete all previous etcd data/config and rejoin the etcd cluster.
2. **By talking to some other node:** If that node did not crash, we should be fine.